README.md
# aws_lambda_s3_python

## Overview
The code will ingest data from Random API into AWS s3 Bucket on 5 minutes interval. The code was ran by Lambda Function that triggered by CloudWatch Events and can see the log on Cloud Watch monitor

## Pre-requisite

Installed [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html#cliv2-linux-install)

```bash
$ curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
$ unzip awscliv2.zip
$ sudo ./aws/install
$ aws configure
$ pip install boto3
```

Install [Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli)
```bash
$ sudo apt-get update && sudo apt-get install -y gnupg software-properties-common
```

Install [boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html)
```bash
$ pip install boto3
```

## Usage

To run this example you need to execute:

```bash
$ terraform init

# without variables
$ terraform plan
$ terraform apply

# with variables
$ terraform plan --var-file="[INPUT_HERE]"
$ terraform apply --var-file="[INPUT_HERE]"
$ terraform output -state="terraform.tfstate"
$ terraform refresh --var-file="[INPUT_HERE]" -lock=false
```



## Assumption
  1. The IAM user that deployed the infrastructure has sufficient permission to apply and make change to the system
  2. The data stored is in json format that will be stored as-is, which will be used for transformation in the downstream process
  3. User has sufficient to create role, iam policy to archieve similar things
  4. This code is not for production purpose, therefore there will be reluctant on certain aspect, such as code format, security concerns, storage
  5. Data store in s3 is unstructure raw data, therefore will be needed further cleaning and transformation to be usable

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.0.3](https://github.com/tuananh8497/aws_lambda_s3_python/releases/tag/v.0.0.3) (2023-05-09)

### Features

* Add output.tf 
* Add variables.tf
* Add versions.tf
* Remove hardcode, use .tfvars


## [0.0.2](https://github.com/tuananh8497/aws_lambda_s3_python/commit/1abaafe92783fdab5579841fc62df897475e9786) (2023-05-07)

### Features

* Add versions.tf 
* Add CHANGELOG.md

## [0.0.1](https://github.com/tuananh8497/aws_lambda_s3_python/commit/1abaafe92783fdab5579841fc62df897475e9786) (2023-05-05)

### Features

* Add first draft of project that meet basic requirement for the technical challenges 
* Current code are running in main.tf and main.py


python/test.py
import boto3
import json
import requests
import logging
import datetime
from botocore.exceptions import ClientError

S3_BUCKET = "bucket-test-20230504222500"
BASE_URL = "https://random-data-api.com/api/v2/users?size=2&response_type=json"


def upload_json(data, bucket_name, object_name=None):
    """Upload a file to an S3 bucket
    :param file_name: File to upload
    :param bucket: Bucket to upload to
    :param object_name: S3 object name. If not specified then file_name is used
    :return: True if file was uploaded, else False
    """
    # Upload the file
    s3_client = boto3.client('s3')
    try:
        data_json = json.dumps(data)
        s3_client.put_object(Bucket=bucket_name,
                             Key=object_name, Body=data_json)
    except ClientError as e:
        logging.error(e)
        return False
    return True


def lambda_handler(event, context):
    response = requests.get(BASE_URL)

    if response.status_code == 200:
        data = response.json()
        print(data)
        for obj in data:
            upload_json(obj, bucket_name=S3_BUCKET, object_name=str(datetime.datetime.utcnow()))
        return {
            'statusCode': 200,
            'body': json.dumps(data)
        }
    else:
        print(f"Error: {response.status_code} - {response.text}")
        return {
            'statusCode': 400,
            'body': json.dumps(f"Error: {response.status_code} - {response.text}"),
        }


print(lambda_handler(0, 0))


python/upload_s3.py
import boto3
import json
import requests
import logging
from botocore.exceptions import ClientError

S3_BUCKET = "bucket-test-20230504222500"
BASE_URL = "https://random-data-api.com/api/v2/users?size=2&response_type=json"

def upload_json(data, bucket_name, object_name=None):
    """Upload a file to an S3 bucket
    :param file_name: File to upload
    :param bucket: Bucket to upload to
    :param object_name: S3 object name. If not specified then file_name is used
    :return: True if file was uploaded, else False
    """
    # Upload the file
    s3_client = boto3.client('s3')
    try:
        data_json = json.dumps(data)
        s3_client.put_object(Bucket=bucket_name, Key=object_name, Body=data_json)
    except ClientError as e:
        logging.error(e)
        return False
    return True

def lambda_handler():
    response = requests.get(BASE_URL)

    if response.status_code == 200:
        data = response.json()
        print(data)
        # upload_json(data, bucket_name=S3_BUCKET, object_name='test_1')
        for obj in data:
            upload_json(obj, bucket_name=S3_BUCKET, object_name=obj['uid'])
        return True
    else:
        print(f"Error: {response.status_code} - {response.text}")
        return False
    
def main():
    lambda_handler()

if __name__ == '__main__':
    main()




python/main.py
import boto3
import json
import requests
import logging
from botocore.exceptions import ClientError
import os 

# S3_BUCKET = "api-staging-bucket-20230505" #actual

S3_BUCKET_NAME = os.environ['S3_BUCKET_NAME']

BASE_URL = "https://random-data-api.com/api/v2/users?size=2&response_type=json"

def upload_json(data, bucket_name, object_name=None):
    """Upload a file to an S3 bucket
    :param file_name: File to upload
    :param bucket: Bucket to upload to
    :param object_name: S3 object name. If not specified then file_name is used
    :return: True if file was uploaded, else False
    """
    # Upload the file
    s3_client = boto3.client('s3')
    try:
        data_json = json.dumps(data)
        s3_client.put_object(Bucket=bucket_name, Key=object_name, Body=data_json)
        print('success put_object')
        print(data_json)
    except ClientError as e:
        logging.error(e)
        return False
    return True

def lambda_handler(event, context):
    # send GET request
    response = requests.get(BASE_URL)

    if response.status_code == 200:
        data = response.json()
        for obj in data:
            upload_json(obj, bucket_name=S3_BUCKET_NAME, object_name=obj['uid'])
        return {
        'statusCode': 200,
        'body': json.dumps(data)
        }
    else:
        print(f"Error: {response.status_code} - {response.text}")
        return {
        'statusCode': 400,
        'body': json.dumps(f"Error: {response.status_code} - {response.text}")
        }
